source/docs/software/networktables/using-messagepack.rst

```diff
@@ -0,0 +1,124 @@
+Using MessagePack
+=================
+
+Passing structured data through Network Tables is now possible using
+`MessagePack <https://msgpack.org/index.html>`__, a format that is also
+understood by the Glass tool, which renders a nice little tree.
+
+Here are a few snippets of how to do it in Java on Windows, which could be used
+to make a little dashboard app that can be run by the Driver Station.  These
+snippets are simplified, eliding several parts required to make them work.
+To run it yourself, check out the complete demo found
+`here.  <https://github.com/truher/radar>`__
+
+There's more than one way to get data in and out of MessagePack; for example,
+the `MessagePack for Java <https://github.com/msgpack/msgpack-java>`__ library
+supports manual packing (packDouble, packInt, packByte, etc) using a
+`MessagePacker <https://github.com/msgpack/msgpack-java/blob/develop/msgpack-core/src/main/java/org/msgpack/core/MessagePacker.java>`__.
+but this manual method seems error-prone and tedious.
+
+Instead, we'll use a **data binding approach,** using the
+`Jackson serializer <https://github.com/msgpack/msgpack-java/blob/develop/msgpack-jackson/README.md>`__.
+Jackson supports many different binding methods; we'll just use POJOs.
+
+First, for our little dashboard, there's the notion of a target and a container.
+Note there are no annotations or anything like that here, Jackson doesn't need
+them for this simple case.
+
+.. tabs::
+
+    .. group-tab:: Java
+
+        .. code-block:: java
+
+            public class Target {
+                public enum Type {
+                    TAG, ALLY, OPPONENT, SELF
```

Comment:
i'll try using the AprilTag class directly and update.

Reply:
ah, i remember why i didn't do that in the first place: (a) i didn't want to deal with pose3d since it's a 2d visualization and (b) i wanted to use the same representation for other kinds of targets, e.g. other robots, the "self" observer, etc.  also i have this tickle in the back of my mind that says you should never use domain objects for IPC, i.e. better to decouple the models of apps that don't get updated or deployed as a unit.